aten/src/ATen/Retainable.h

```diff
@@ -7,21 +7,52 @@ namespace at {
 // base class for refcounted things, allows for collects of generic
 // refcounted objects that include tensors
 struct Retainable {
-  Retainable(): refcount(1) {}
+  Retainable(): refcount(1), weak_refcount(1) {}
```

Comment:
would it cause perf overhead as Retainable is used in several places and it increases the size of struct?

Maybe I'm missing the context, but what is the motivation for this PR? It diverges quite a bit from the unified design we were hashing out. @ezyang @smessmer

Reply:
The immediate motivation for weak references is for thread-local tracing (https://github.com/pytorch/pytorch/pull/9414). That in and of itself wouldn't be a good reason, but we had a previous case where we needed weak references for a weak mapping of storages to file descriptors in our multiprocessing implementation. And we think that weak references are generally useful enough to deserve the extra word in the tensor struct to hold them.

Reply:
Also, if it really turns out to be too expensive for some reason, we could effectively pack the weak refcount together with regular refcount in a single integer (but I'd really treat that as a last resort).